Feature grpc server

prravda opened the pull request.

prravda committed:
impl:
- created .proto file for defining gRPC and protobuf

prravda committed:
impl:
- implemented a simple gRPC server code

prravda committed:
impl:
- try to wrap up functions with asyncio

prravda committed:
impl
- detach the tokenizer code from protobuf servicer
- add comments for getting help

prravda commented on src/tokenizing.py:
```diff
@@ -10,28 +12,61 @@
 
-async def tokenize(sentence: tokenizer.TokenizingRequest) -> tokenizer.TokenizingResult:
+@dataclass
+class TokensAndWords:
```
@Hyunhyun12
- 이전 코드가 gRPC 에서 사용하는 타입이 형태소 분석 함수에도 얽혀져 있어서 일단 분리하였습니다.

prravda commented on src/tokenizing.py:
```diff
@@ -10,28 +12,61 @@
+    tokens: list[str]
+
+
+async def tokenize(title: str) -> TokensAndWords:
```
@Hyunhyun12
- 형태소 분석을 하는 코드입니다.
- 라이브러리에서 `asyncio` 는 지원하지 않는 상황이구요, asyncio 의 `event_loop` 기능으로 wrapping 만 해 놨습니다. 
- 이 부분은 `gRPC` 와는 무관한 부분입니다.

Hyunhyun12 replied:
라이브러리에서 asyncio 는 지원하지 않는다면 코루틴 내 비동기 과정이 없다는 의미입니다. 그렇다면 비동기식을 사용하는 의미가 없습니다. 그래서 제가 키위 라이브러리 문서를 읽어봤는데, 이미 사용 가능한 모든 스레드가 동원되어 있더라고요. 라이브러리를 정말 잘 만들어 놨습니다.

https://bab2min.github.io/kiwipiepy/v0.15.2/kr/
위 링크에서 
"멀티스레딩 analyze
다음 예제 코드는 멀티스레드를 활용하여 test.txt 파일을 줄별로 읽어들여 형태소 분석한 뒤 그 결과를 result.txt 에 저장합니다."
부분을 보시면 됩니다.

Hyunhyun12 replied:
아 혹시 서버 io 부분을 비동기화하길 원하시나요? 그렇다면 방법이 있습니다.
일단 asyncio 모듈은 고수준 API와 저수준 API가 있는데, 그 중 본 파일에서 사용하신 loop는 저수준 API입니다. 아래 코드에서 계속 설명드릴게요.

prravda commented on src/tokenizing.py:
```diff
@@ -10,28 +12,61 @@
+    return TokensAndWords(tokens, words)
 
 
 class TokenizerServicer(tokenizer_grpc.TokenizerServicer):
```
@Hyunhyun12
- 여기서부터가 protobuf code 입니다. 
- 형태소 분석을 처리하는 `tokenize` 함수를 `await` 으로 실행하고, 그 결과를 `.proto` 파일로 만든 response type 객체에 mapping 합니다.

prravda commented on src/tokenizing.py:
```diff
@@ -10,28 +12,61 @@
 
 
-async def serve():
+async def serve() -> None:
```
@Hyunhyun12
- 이 부분은 서버를 실행시키는 코드입니다. 
- `grpc.aio.server` 를 통해 async server 를 실행합니다.

prravda commented:
@Hyunhyun12
- python 의 모던한 비동기 처리에 대해서 조언을 구하고자 호출합니다! 

궁금한 점입니다.
- asyncio 를 대응하지 않는 형태소분석기 코드를 `async` 로 wrapping 하는 방식
  - 이건 제가 직접 찾아보면서 작성하였습니다. 
- gRPC 서버를 asyncio 로 구동하는 방식 
  - 이건 gRPC 에서 제공한 (asynchronous 하게 gRPC 서버를 구동하는 코드)[https://github.com/grpc/grpc/blob/v1.57.0/examples/python/route_guide/asyncio_route_guide_server.py#L135] 를 참고하였습니다.

Hyunhyun12 reviewed (commented):
black & isort

Hyunhyun12 reviewed (commented):
black & isort 일단 반영하였습니다

Hyunhyun12 commented:
아 오케이!! 어떻게 쓰는지 알았어요

Hyunhyun12 commented:
일단 전체적으로..

1. python 3.11에서 성능개선 및 asyncio 모듈 패러다임 변경이 이루어졌으므로 3.11 추천드립니다. conda python도 3.11이 되었으니 충분히 안정화됐다고 생각해요 (3.10에서 3.11로 가는 업데이트는 사실 크게 문법이 변한 건 없고 파이썬 내부의 C언어 최적화가 잘 되면서 성능이 극적으로 향상됐습니다) 저는 miniconda python 3.11을 씁니다. 가볍고 좋아요.


2. 파이썬의 선(Zen of Python)을 보시면 https://wikidocs.net/7907 아래와 같은 문장이 있습니다.

=========================================================================

Beautiful is better than ugly. 아름다움이 추한 것보다 낫다.

Explicit is better than implicit. 명확함이 함축된 것보다 낫다.

Simple is better than complex. 단순함이 복잡한 것보다 낫다.

Complex is better than complicated. 복잡함이 난해한 것보다 낫다.

Flat is better than nested. 단조로움이 중접된 것보다 낫다.

Sparse is better than dense. 여유로움이 밀집된 것보다 낫다.

Readability counts. 가독성은 중요하다.

Special cases aren't special enough to break the rules. 규칙을 깨야할 정도로 특별한 경우란 없다.

Although practicality beats purity. 비록 실용성이 이상을 능가한다 하더라도.

There should be one -- and preferably only one -- obvious way to do it. 문제를 해결할 하나의 - 바람직하고 유일한 - 명백한 방법이 있을 것이다.

=========================================================================

특히 가장 마지막에 소개해드린 "문제를 해결할 하나의 - 바람직하고 유일한 - 명백한 방법이 있을 것이다" 부분이 파이썬에서는 중요한데, 바로 코드 스타일을 표준화했다는 점입니다. 코드 스타일은 C/C++에서는 소위 "종교적 논쟁"이라고 불리며 이에 대해 논쟁하는 것을 금기시합니다. 그러나 파이썬 재단은 이를 표준화시켰고, 코드를 표준 코드 스타일로 다시 써주는 black이라는 라이브러리를 직접 관리하고 있습니다. 또한 import문을 서술하는 표준도 있어서, isort라는 라이브러리를 통해 관리할 수 있습니다. 저는 conda install black, conda install isort 한 후에, black .와 isort .를 수시로 실행시켜 줍니다. 또한 사용하지 않는 import문을 자동으로 제거해 주는 autoflake라는 라이브러리도 커밋 전에 사용합니다.

Hyunhyun12 commented:
또한 폴더마다 있는 __init__.py는 파이썬 3.2까지는 필수적이었으나, 3.3부터는 필수적이지 않습니다. __init__.py가 있으면 regular package, 없으면 namespace package라고 불립니다. 한국어 공식문서는 아래에 있는데 번역이 별로입니다.
https://docs.python.org/ko/3/reference/import.html#packages

영문 공식문서는 아래에 있습니다.
https://docs.python.org/3/reference/import.html#packages

요약하면, 지금 protos 폴더나 src 폴더는 전체 파일의 일부분이므로, namespace package를 사용하는 것이 맞고 그러므로 폴더 안에 있는 __init__.py는 제거되어야 합니다. (제거하지 않아도 실행에는 문제가 없습니다)

Hyunhyun12 commented on app.py:
```diff
@@ -1,8 +1,17 @@
 from kiwipiepy import Kiwi
+import tokenizer_definition_pb2
```
파일 하나를 전부 와일드카드 임포트하고 있습니다. 잘 격리하려면
from src.proto_built_result.tokenizer_definition_pb2 import TokenizingResult
와 같이 가져오시면 됩니다. 파이썬은 함수 또한 일급 객체이므로 이렇게 사용하셔도 코드가 작동합니다.
(그리고 더 빠릅니다. 와일드카드 임포트를 했을 때는 인터프리터가 파일 내의 여러 객체 중 TokenizingResult()를 찾는 시간이 추가로 걸리기 때문입니다.)

Hyunhyun12 commented on app.py:
```diff
@@ -1,8 +1,17 @@
+
 kiwi = Kiwi()
 
+result = tokenizer_definition_pb2.TokenizingResult()
```
from src.proto_built_result.tokenizer_definition_pb2 import TokenizingResult
로 가져오셨다면, result = TokenizingResult()로 받으시면 됩니다.

Hyunhyun12 commented on app.py:
```diff
@@ -1,8 +1,17 @@
+
 hotdeal_title = "[지마켓] 가시제거연구소 노르웨이 순살고등어 오렌지라벨 800g+800g (20,630원) (무료)"
-result = kiwi.tokenize(hotdeal_title)
+tokenizer_result = kiwi.tokenize(hotdeal_title)
```
List[Token] 형태의 값이 저장되는 것으로 보입니다. 그렇다면 from typing import List 및 Token에 관한 import statement를 파일 위에 넣어두고, tokenizer_result: List[Token] = kiwi.tokenize(hotdeal_title) 을 하는 방법으로 제네릭을 사용하실 수 있습니다. List 대소문자 구분에 유의해주세요!! list로 쓰시면 안됩니다!!

Hyunhyun12 commented on app.py:
```diff
@@ -1,8 +1,17 @@
-result = kiwi.tokenize(hotdeal_title)
+tokenizer_result = kiwi.tokenize(hotdeal_title)
+
+extracted_tokens = []
```
11-13번 라인에는 리스트 컴프리헨션을 사용하시면 간단하고 속도도 약 3배정도 빠릅니다. 제네릭도 사용하면
extracted_tokens: List[str] = [t.form for t in tokenizer_result]

Hyunhyun12 commented on app.py:
```diff
@@ -1,8 +1,17 @@
+for t in tokenizer_result:
+    extracted_tokens.append(t.form)
+
+result.tokens.extend(extracted_tokens)
```
result는 파이썬에서 함수의 리턴값을 관용적으로 가리키는 변수명이라서, 후일 함수 안에 넣거나 또는 바꾸는 것이 좋아 보입니다.

Hyunhyun12 commented on src/tokenizing.py:
```diff
@@ -0,0 +1,87 @@
+import grpc
+import asyncio
+
+from dataclasses import dataclass
```
dataclasses는 파이썬 기본 모듈로 들어가 있지만, 실무에서는 잘 사용되지 않는 것 같습니다. 왜냐하면 dataclasses의 기능(자료구조를 간편하게 정의할 수 있다) 및 파싱 기능까지 포함되어 있는 pydantic이라는 라이브러리가 있기 때문입니다. 예를 들어, dataclasses에는 str 자료형에 None(널포인터)를 넣으면 에러가 뜨고 서버가 터질 수 있지만, pydantic에서는 자동으로 ""(빈 문자열)로 파싱해 줍니다.

Hyunhyun12 commented on src/tokenizing.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    defining the result of tokenizing
+    """
+    words: list[str]
```
from typing import List 로 임포트하신 후, List[str]로 자료형을 정의하시면 됩니다. app.py 파일에서도 말씀드렸듯이, list와 List 대소문자 구분 중요합니다!

Hyunhyun12 commented on src/tokenizing.py:
```diff
@@ -0,0 +1,87 @@
+    - cf. the tokenizer library(Kiwi) is not supports asyncio itself
+    """
+
+    loop = asyncio.get_event_loop()
```
음... https://docs.python.org/ko/3/library/asyncio-eventloop.html 이 문서를 보면,"응용 프로그램 개발자는 일반적으로 asyncio.run()과 같은 고수준의 asyncio 함수를 사용해야 하며, 루프 객체를 참조하거나 메서드를 호출할 필요가 거의 없습니다. 이 절은 주로 이벤트 루프 동작을 세부적으로 제어해야 하는 저수준 코드, 라이브러리 및 프레임워크의 작성자를 대상으로 합니다."라고 되어 있습니다. 그래서 저는 저수준 API를 사용해 본 적이 없지만, 공부한 것을 말씀드려 보겠습니다.

Hyunhyun12 commented on src/tokenizing.py:
```diff
@@ -0,0 +1,87 @@
+    loop = asyncio.get_event_loop()
+
+    with futures.ThreadPoolExecutor() as pool:
+        parsed_tokens = await loop.run_in_executor(pool, kiwi_instance.tokenize, title.title)
```
이 부분도 권장되지 않는 스타일입니다. *arg부분은 kiwi_instance.tokenize(title.title)로 들어갈 부분으로 보이는데, 함수를 일급 객체로 다루는 파이썬에서는 run_in_executor(pool, kiwi_instance.tokenize, title.title)처럼 코드를 쓰는 것은 권장되지 않습니다. (제가 느끼기에는 C/C++ 문법으로 파이썬 코드를 쓰는 것 같습니다) 대신 파이썬에서는 asyncio 고수준 API의 Task 객체를 활용합니다. 코루틴(비동기함수)를 Task의 형태로 감싸서 한번 예약한 후, 예약된 Task를 TaskGroup으로 만들어서 한꺼번에 run으로 돌려버리거나, 여러개의 TaskGroup을 관리하거나, 새로운 Task를 기존의 이벤트루프에 추가하거나 합니다.

Hyunhyun12 replied:
또한 타입 힌트가 지정되지 않아서 List[Token] | Iterable[List[Token]] | List[List[Token]] | Iterable[List[List[Token]]]
4개 중 하나로 되어 있습니다. List[Token]으로 보이니, parsed_tokens: List[Token] = await loop.run_in_executor(pool, kiwi_instance.tokenize, title.title) 으로 작성하시면 됩니다.

Hyunhyun12 commented on src/tokenizing.py:
```diff
@@ -0,0 +1,87 @@
+    with futures.ThreadPoolExecutor() as pool:
+        parsed_tokens = await loop.run_in_executor(pool, kiwi_instance.tokenize, title.title)
+
+    tokens: list[str] = []
```
app.py에서와 같이, 리스트 컴프리헨션을 사용하시면 코드도 더 간략하고 속도도 더 빠릅니다.

Hyunhyun12 commented on src/tokenizing.py:
```diff
@@ -0,0 +1,87 @@
+      which generated by protoc compiler using .proto file
+    """
+
+    async def Tokenize(self, request: tokenizer.TokenizingRequest, unused_context) -> tokenizer.TokenizingResult:
```
PEP 8에 위배되는 함수명입니다. 작동에는 문제가 없지만, 권장되지 않습니다. 함수명은 snake_case가 권장됩니다.

prravda replied:
이건 protobuf 의 naming convention  의 관행인 듯 하여 이렇게 작성하였습니다.

prravda commented:
방금까지 작업을 하다 지금 확인하였습니다.
진짜 많이 배워갑니다. 

감사합니다. 

- pydantic 적용
- 기본적인 linter 구성
- 형태소 분석 모델로부터 async wrapping 제거
- 형태소 분석 결과로부터 문자열 추출하는 과정을 list comprehension 처리 
- zen of python 읽어보고, pythonic 한 코드로 작성

해 보도록 하겠습니다!

Hyunhyun12 commented on src/tokenizing.py:
```diff
@@ -0,0 +1,87 @@
+    - then, add servicer to server
+    - finally, open up the port and running the server
+    """
+    server = grpc.aio.server(futures.ThreadPoolExecutor())
```
아 여기서 저수준 API를 사용해서 전체적으로 저수준 API를 사용하셨군요. asyncio와 concurrent.futures를 섞어 쓰는 건 GIL 때문에 성능 향상이 없고 오히려 오버헤드가 발생하여 성능을 낮출 우려가 있다고 합니다. https://stackoverflow.com/questions/69728812/asyncio-and-concurrent-futures-threadpoolexecutor

Hyunhyun12 replied:
https://www.reddit.com/r/Python/comments/m7sf7y/asyncio_and_concurrentfutures/
이 부분의 2번째 답변을 보면

concurrent.futures offers a convenient API for doing work with either threads or processes. I don't really use asyncio, but I understand the equivalent is asyncio.Future.

I expect that the reason it wasn't added to the concurrent library is because of the await keyword. You can't simply mix code that uses await with code that doesn't. That's why so many libraries are being completely rewritten to work with asyncio.

To answer your main question, oversimplified a bit:

If you are doing a lot of heavy processing work, use multiprocessing (or concurrent.futures.ProcessPoolExecutor).

If you are doing occasional IO, or background work for a GUI (less than 50 tasks or so), use threading (or concurrent.futures.ThreadPoolExecutor).

If you are doing heavy IO (100+ tasks), then it makes sense to use asyncio. There is controversy over asyncio so there are alternatives, but they work fundamentally the same, just with different interfaces. Some argue that you're better off with curio or gevent (my preference, does not use await).

Hyunhyun12 replied:
라고 합니다. 음... 저는 asyncio를 추천드려요. 왜냐하면 파이썬 재단이 지원도 많이 해주고 있고, 3.11이 되면서 Go statement considered harmful 패러다임도 도입되어 있어서 아키텍처도 깔끔하게 만들 수 있거든요. 하지만 이건 제가 결정할 부분이 아니라서, 뜻대로 하시면 될 것 같습니다.

https://grpc.github.io/grpc/python/grpc_asyncio.html#

이렇게 grpc에서 공식 예시를 제공할 정도입니다!

prravda replied:
[gRPC 의 해당 부분](https://grpc.github.io/grpc/python/grpc_asyncio.html#create-server) 에서 `migration_thread_pool` 에 들어가는 parameter type 이 `futures.ThreadPoolExecutor` 라 확인한 gRPC 예제 코드에서 아직 더 이상의 업데이트는 없는 듯 합니다.

Hyunhyun12 commented on Dockerfile:
```diff
@@ -1,5 +1,4 @@
 FROM python:3.10
```
3.11 진짜 좋아요

prravda committed:
chore
- remove `__init__.py` files, because they are parts of current system, so they are namespace package

prravda committed:
chore
- removed unused file

prravda committed:
chore
- update python image from 3.10 to 3.11

prravda committed:
dependencies
- installed black, pydantic, and isort

prravda committed:
linting
- used `black` linter and formatted these files

prravda committed:
impl:
- created a new `docker-compose.yaml` file

prravda commented:
- dataclass → pydantic
- linter configuration (isort, black)
- updated to python 3.11 
- created a new docker-compose.yaml file
- removed useless async wrapping
